File: java/easy/minimum_waiting_time/MinimumWaitingTime.java
package main.java.com.example.minimumwaitingtime;

import java.util.Arrays;

public class MinimumWaitingTime {
	// Function to find the minimum waiting time for all of the queries in an array.
	// O(nlog(n)) time | O(1) space, where n is the total number of queries
	public static int minimumWaitingTime(int[] queries) {
		// Sort the queries to minimize the waiting time.
		Arrays.sort(queries);

		int totalWaitingTime = 0;
		for(int idx = 0; idx < queries.length; idx++) {
			// Calculate how many queries are left after the current one.
			int queriesLeft = queries.length - (idx + 1);
			// Accumulate the waiting time contribution of the current query.
			totalWaitingTime += queries[idx] * queriesLeft;
		}

		return totalWaitingTime;
	}
}

File: python/easy/minimum_waiting_time/minimum_waiting_time.py
# Function to find the minimum waiting time for all of the queries in an array.
# O(nlog(n)) time | O(1) space, where n is the total number of queries
def minimumWaitingTime(queries):
    # Sort the queries in ascending order to minimize the waiting time.
    queries.sort()

    totalWaitingTime = 0
    for idx, duration in enumerate(queries):
        # Calculate how many queries are left after the current one.
        queriesLeft = len(queries) - (idx + 1)
        # Accumulate the waiting time contribution of the current query.
        totalWaitingTime += queries[idx] * queriesLeft

    return totalWaitingTime